File: javascript/libjoynr-js/src/main/js/joynr/dispatching/types/SubscriptionRequest.js
/*jslint node: true */

/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
var Util = require("../../util/UtilInternal");
var Typing = require("../../util/Typing");
var PeriodicSubscriptionQos = require("../../proxy/PeriodicSubscriptionQos");
var OnChangeSubscriptionQos = require("../../proxy/OnChangeSubscriptionQos");
var OnChangeWithKeepAliveSubscriptionQos = require("../../proxy/OnChangeWithKeepAliveSubscriptionQos");

var defaultSettings = {
    qos: new PeriodicSubscriptionQos()
};

/**
 * @name SubscriptionRequest
 * @constructor
 * @param {Object} settings
 * @param {String}
 *            settings.subscriptionId Id of the new subscription
 * @param {String}
 *            settings.subscribedToName the name of the element to subscribe to
 * @param {SubscriptionQos}
 *            [settings.qos] the subscriptionQos
 */
function SubscriptionRequest(settings) {
    Typing.checkProperty(settings, "Object", "settings");
    Typing.checkProperty(settings.subscriptionId, "String", "settings.subscriptionId");
    Typing.checkProperty(settings.subscribedToName, "String", "settings.subscribedToName");
    Typing.checkPropertyIfDefined(
        settings.qos,
        ["Object", "OnChangeSubscriptionQos", "PeriodicSubscriptionQos", "OnChangeWithKeepAliveSubscriptionQos"],
        "settings.qos"
    );

    /**
     * @name SubscriptionRequest#subscriptionId
     * @type String
     */
    /**
     * @name SubscriptionRequest#qosName
     * @type OnChangeSubscriptionQos|PeriodicSubscriptionQos|OnChangeWithKeepAliveSubscriptionQos
     */
    /**
     * @name SubscriptionRequest#attributeName
     * @type String
     */
    Util.extend(this, defaultSettings, settings);

    /**
     * The joynr type name
     *
     * @name SubscriptionRequest#_typeName
     * @type String
     */

    Object.defineProperty(this, "_typeName", {
        value: "joynr.SubscriptionRequest",
        readable: true,
        writable: false,
        enumerable: true,
        configurable: false
    });

    return Object.freeze(this);
}

/**
 * The joynr type name
 *
 * @name Request#_typeName
 * @type String
 */
Object.defineProperty(SubscriptionRequest, "_typeName", {
    value: "joynr.SubscriptionRequest",
    readable: true,
    writable: false,
    enumerable: true,
    configurable: false
});

module.exports = SubscriptionRequest;

File: cpp/libjoynr/basemodel/subscription/SubscriptionRequest.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
#include "joynr/SubscriptionRequest.h"

#include "joynr/SubscriptionQos.h"
#include "joynr/Util.h"

namespace joynr
{

SubscriptionRequest::SubscriptionRequest() : subscriptionId(), subscribedToName(), qos()
{
    subscriptionId = util::createUuid();
}

const std::string& SubscriptionRequest::getSubscriptionId() const
{
    return subscriptionId;
}

const std::string& SubscriptionRequest::getSubscribeToName() const
{
    return subscribedToName;
}

bool SubscriptionRequest::operator==(const SubscriptionRequest& subscriptionRequest) const
{
    bool equal = *qos == *(subscriptionRequest.qos);
    return subscriptionId == subscriptionRequest.getSubscriptionId() &&
           subscribedToName == subscriptionRequest.getSubscribeToName() && equal;
}

void SubscriptionRequest::setSubscriptionId(const std::string& idLocal)
{
    this->subscriptionId = idLocal;
}

void SubscriptionRequest::setSubscriptionId(std::string&& idLocal)
{
    this->subscriptionId = std::move(idLocal);
}

void SubscriptionRequest::setSubscribeToName(const std::string& subscribedToNameLocal)
{
    this->subscribedToName = subscribedToNameLocal;
}

void SubscriptionRequest::setSubscribeToName(std::string&& subscribedToNameLocal)
{
    this->subscribedToName = std::move(subscribedToNameLocal);
}

std::string SubscriptionRequest::toString() const
{
    return joynr::serializer::serializeToJson(*this);
}

std::shared_ptr<SubscriptionQos> SubscriptionRequest::getQos() const
{
    return qos;
}

void SubscriptionRequest::setQos(std::shared_ptr<SubscriptionQos> qosLocal)
{
    this->qos = std::move(qosLocal);
}

} // namespace joynr
